src/domain/entities/Questionnaire/QuestionnaireRules.ts

```diff
@@ -0,0 +1,269 @@
+
+const D2_FUNCTIONS = ["d2:hasValue", "d2:daysBetween", "d2:yearsBetween"];
+const D2_OPERATORS = [
+    ">" as const,
+    ">=" as const,
+    "<" as const,
+    "<=" as const,
+    "==" as const,
+    "!=" as const,
+];
+
+export type QuestionnaireRuleActionType =
+    | "DISPLAYTEXT"
+    | "DISPLAYKEYVALUEPAIR"
+    | "HIDEFIELD"
+    | "HIDESECTION"
+    | "ASSIGN"
+    | "SHOWWARNING"
+    | "SHOWERROR"
+    | "WARNINGONCOMPLETINON"
+    | "ERRORONCOMPLETION"
+    | "CREATEEVENT"
+    | "SETMANDATORYFIELD";
+
+export interface QuestionnaireRuleAction {
+    id: Id;
+    programRuleActionType: QuestionnaireRuleActionType;
+    dataElement?: {
+        id: Id | undefined; // to hide
+    };
+    data?: string; // to assign
+    programStageSection?: {
+        id: Id | undefined; // to hide/show
+    };
+    programStage?: {
+        id: Id | undefined; // to hide/show
+    };
+    content?: string; // message content to show
+}
+export interface QuestionnaireRule {
+    id: Id;
+    condition: string; //condition is parsed with dataelementId e.g: #{dataElementId} == 'Yes'
+    dataElementIds: Id[]; // all dataElements in condition (there could be mutiple conditions)
+    actions: QuestionnaireRuleAction[];
+    parsedResult?: boolean; //calculate the condition and store the result
+}
+
+export const getApplicableRules = (
+    updatedQuestion: Question,
+    questionnaireRules: QuestionnaireRule[],
+    questions: Question[]
+): QuestionnaireRule[] => {
+    //1. Get all Rules that are applicable to the updated question
+    const applicableRules = questionnaireRules.filter(
+        rule =>
+            rule.dataElementIds.includes(updatedQuestion.id) ||
+            rule.actions.some(action => action.dataElement?.id === updatedQuestion.id)
+    );
+
+    //2. Run the rule conditions and return rules with parsed results
+    const parsedApplicableRules = applicableRules.map(rule => {
+        const parsedResult = parseCondition(rule.condition, updatedQuestion, questions);
+        return { ...rule, parsedResult };
+    });
+
+    return parsedApplicableRules;
+};
+
+const getQuestionValueByType = (question: Question): string => {
+    switch (question.type) {
+        case "select":
+            return question.value?.code ?? "";
+        case "boolean":
+            return question.value === undefined ? "false" : question.value.toString();
+        case "date":
+        case "datetime":
+            return question.value?.toString() ?? "";
+
+        case "number":
+        case "text":
+            return question.value ?? "";
+        default:
+            console.debug("Unknown question type");
+            return "";
+    }
+};
+
+const parseConditionValues = (
+    condition: string,
+    updatedQuestion: Question,
+    questions: Question[]
+) => {
+    return condition.replace(/#\{(.*?)\}/g, (_i, dataElementId) => {
+        //first check if the dataElementId is in the updatedQuestion,
```

Comment:
[general comment] The comment seems a bit redundant, but in any case it can be transformed to code: `const isDataElementInUpdatedQuestion = updatedQuestion.id === dataElementId` and later `if (isDataElementInUpdatedQuestion)`